scipy/linalg/tests/test_lapack.py

```diff
@@ -1645,3 +1645,134 @@ def test_getc2_gesc2():
+    # non zero info.
+
+    np.random.seed(42)
+    n = 10
+    rtol = 250*np.finfo(dtype).eps  # set test tolerance appropriate for dtype
+
+    # create the matrix in accordance with the data type
+    du = my_rand(n-1, dtype=dtype)
+    d = my_rand(n, dtype=dtype)
+    dl = my_rand(n-1, dtype=dtype)
+
+    diag_cpy = [dl.copy(), d.copy(), du.copy()]
+
+    A = np.diag(d) + np.diag(dl, -1) + np.diag(du, 1)
+    x = np.random.rand(n)
+    b = A @ x
+
+    gttrf, gttrs = get_lapack_funcs(('gttrf', 'gttrs'), dtype=dtype)
+
+    _dl, _d, _du, du2, ipiv, info = gttrf(dl, d, du)
+    # test to assure that the inputs of ?gttrf are unmodified
+    assert_array_equal(dl, diag_cpy[0])
+    assert_array_equal(d, diag_cpy[1])
+    assert_array_equal(du, diag_cpy[2])
+
+    # generate L and U factors from ?gttrf return values
+    # L/U are lower/upper triangular by construction (initially and at end)
+    U = np.diag(_d, 0) + np.diag(_du, 1) + np.diag(du2, 2)
+    L = np.eye(n, dtype=dtype)
+
+    for i, m in enumerate(_dl):
+        # L is given in a factored form.
+        # See http://www.hpcavf.uclan.ac.uk/softwaredoc/sgi_scsl_html/sgi_html/ch03.html
+        piv = ipiv[i] - 1
+        # right multiply by permutation matrix
+        L[:, [i, piv]] = L[:, [piv, i]]
+        # right multiply by Li, rank-one modification of identity
+        L[:, i] += L[:, i+1]*m
+
+    # one last permutation
+    i, piv = -1, ipiv[-1] - 1
+    # right multiply by final permutation matrix
+    L[:, [i, piv]] = L[:, [piv, i]]
+
+    _A = L @ U
+    # check that the outputs of ?gttrf define an LU decomposition of A
+    assert_allclose(A, _A, atol=rtol)
+
+    b_cpy = b.copy()
+    x_gttrs, info = gttrs(_dl, _d, _du, du2, ipiv, b)
+    # test that the inputs of ?gttrs are unmodified
+    assert_array_equal(b, b_cpy)
+    # test that the result of ?gttrs matches the expected input
+    assert_allclose(x, x_gttrs, rtol=rtol)
+
+    # test that ?gttrf and ?gttrs work with transposal options
+    if dtype in REAL_DTYPES:
+        trans = "T"
+        b_trans = A.T @ x
+    else:
+        trans = "C"
+        b_trans = A.conj().T @ x
+
+    x_gttrs, info = gttrs(_dl, _d, _du, du2, ipiv, b_trans, trans=trans)
+    assert_allclose(x, x_gttrs, rtol=rtol)
+
+    # test that ValueError is raised with incompatible matrix shapes
+    with assert_raises(ValueError):
+        gttrf(dl[:-1], d, du)
+    with assert_raises(ValueError):
+        gttrf(dl, d[:-1], du)
+    with assert_raises(ValueError):
+        gttrf(dl, d, du[:-1])
+
+    # test that matrix of size n=2 raises exception
+    with assert_raises(Exception):
+        gttrf(dl[0], d[:1], du[0])
+
+    # test that singular (row of all zeroes) matrix fails via info
+    du[0] = 0
+    d[0] = 0
+    __dl, __d, __du, _du2, _ipiv, _info = gttrf(dl, d, du)
+    np.testing.assert_(_info != 0,
+                       "?gttrf: singular matrix ret info == 0")
```

Comment:
This test could be slightly more specific. The [?gttrf docs](http://www.netlib.org/lapack/explore-html/d1/d40/group__double_g_tcomputational_ga8ca64e542924cec56cbe9837b77d25b7.html) give the following description for ``info > 0``:

> if INFO = k, U(k,k) is exactly zero. The factorization has been completed, but the factor U is exactly             singular, and division by zero will occur if it is used to solve a system of equations.

Rather than the more general ``_info != 0`` we expect the first element (as you have set du[0] to exactly zero) to be zero, i.e. ``_info == 1``. The reason this is 1 is because we are switching between Fortran and C style indexing.

Reply:
OK.

Reply:
So for this it looks like _info is not giving 1 for the `i`th element, but instead is `info = 10`, so not the first element of the matrix, but the very last?. (it gives 9 when n=9, 8 when n=8, etc).  Is this correct behavior?

Reply:
What is the value of `d[info - 1]` after the factorisation is complete? It could be that it has reordered the factors?

Reply:
I’ll have to come back to check on that tomorrow, but I committed the other changes for now!

Reply:
It looks like the last element of d, d[info-1], is zero.

Reply:
So, does this mean that should implement the test to check that the d[info-1]th value contains the illegal value, zero, that we put in d?

Reply:
Yes. After factorization, `d[k]` is `U[k, k]`. So Fortran (1-indexed) says `k=10`, look in `d[info-1]` for the zero.